narwhal/primary/src/primary.rs

```diff
@@ -555,11 +767,57 @@ impl PrimaryToPrimary for PrimaryReceiverHandler {
         &self,
         request: anemo::Request<PrimaryMessage>,
     ) -> Result<anemo::Response<()>, anemo::rpc::Status> {
-        let message = request.into_body();
-        self.tx_primary_messages
-            .try_send(message)
-            .map(|_| anemo::Response::new(()))
-            .map_err(|e| anemo::rpc::Status::internal(e.to_string()))
+        let PrimaryMessage::Certificate(certificate) = request.into_body();
+        let (tx_ack, rx_ack) = oneshot::channel();
+        self.tx_certificates
+            .send((certificate, Some(tx_ack))).await
+            .map_err(|e| anemo::rpc::Status::internal(e.to_string()))?;
+        rx_ack.await
+            .map_err(|e| anemo::rpc::Status::internal(e.to_string()))?
+            .map_err(|e| anemo::rpc::Status::internal(e.to_string()))?;
+        Ok(anemo::Response::new(()))
+    }
+
+    async fn request_vote(
+        &self,
+        request: anemo::Request<RequestVoteRequest>,
+    ) -> Result<anemo::Response<RequestVoteResponse>, anemo::rpc::Status> {
+        // TODO: Remove manual code for tracking inflight requests once Anemo issue #9 is resolved.
+        let author = request.body().header.author.to_owned();
+        let _inflight_guard = if self.request_vote_inflight.insert(author.clone()) {
+            RequestVoteInflightGuard {
+                request_vote_inflight: self.request_vote_inflight.clone(),
+                author,
+            }
+        } else {
+            return Err(anemo::rpc::Status::new_with_message(
+                // TODO: This should be 429 Too Many Requests, if/when Anemo adds that status code.
```

Comment:
nit: might be worth opening an anemo fr for this so its tracked?

Reply:
Filed https://github.com/MystenLabs/anemo/issues/13 - @bmwill